editor/engine_update_label.cpp

```diff
@@ -0,0 +1,344 @@
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "engine_update_label.h"
+
+#include "core/os/time.h"
+#include "editor/editor_settings.h"
+#include "editor/editor_string_names.h"
+#include "editor/themes/editor_scale.h"
+#include "scene/gui/box_container.h"
+#include "scene/gui/button.h"
+#include "scene/main/http_request.h"
+
+bool EngineUpdateLabel::_can_check_updates() const {
+	return int(EDITOR_GET("network/connection/network_mode")) == EditorSettings::NETWORK_ONLINE &&
+			UpdateMode(int(EDITOR_GET("network/connection/engine_version_update_mode"))) != UpdateMode::DISABLED;
+}
+
+void EngineUpdateLabel::_check_update() {
+	checked_update = true;
+	_set_status(UpdateStatus::BUSY);
+	http->request("https://raw.githubusercontent.com/godotengine/godot-website/master/_data/versions.yml");
+}
+
+void EngineUpdateLabel::_http_request_completed(int p_result, int p_response_code, const PackedStringArray &p_headers, const PackedByteArray &p_body) {
+	if (p_result != OK) {
+		_set_status(UpdateStatus::ERROR);
+		_set_message(vformat(TTR("Failed to check for updates. Error: %d."), p_result), theme_cache.error_color);
+		return;
+	}
+
+	if (p_response_code != 200) {
+		_set_status(UpdateStatus::ERROR);
+		_set_message(vformat(TTR("Failed to check for updates. Response code: %d."), p_response_code), theme_cache.error_color);
+		return;
+	}
```

Comment:
Just thought about this, but I wonder if we should future proof in case we ever refactor the website and no longer have https://raw.githubusercontent.com/godotengine/godot-website/master/_data/versions.yml

Maybe we should just print verbose these messages, and not show anything in the PM when requests fail?

So people can keep using e.g. Godot 4.3-stable without being nagged if in a year the website has been ported to a new framework and this file no longer exists.

Otherwise, we need to make sure the file we use for these checks will always be online (but that's hard to future proof).

WDYT @Calinou @coppolaemilio?

Reply:
This is a valid concern because we might lose control of githubusercontent.com. However it can be debated we can also lose control of godotengine.org.

Reply:
Yeah I was thinking the same.

We currently add .json files to https://godotengine.org/mirrorlist/ (e.g. https://godotengine.org/mirrorlist/4.2.1.stable.json) for each release, but it's not the most convenient to parse.

It should be fairly easy to add a build step for the website to generate e.g. https://godotengine.org/mirrorlist/versions.json that would include all versions with their respective blog post.

Reply:
Yes, we shouldn't use this file. I would go with our website. We can serve a page like versions.yml in a domain that we control so that if we change servers or things like that we still have the same response from the same address. 

This could be `versions.godotengine.org/yml` or `godotengine.org/versions.yml`. Anything like that will be better.

I can make the changes needed for getting that address up.

Reply:
BTW when we implement that, the description I amended in https://github.com/godotengine/godot/pull/75916/files#diff-52f0c165415d19abd2d5586c8a168169a8d39716be86688a9117125409b421f0R878 should be changed to remove explicit mention of GitHub.